Add auto-certs-renewal design doc (#5687)
aws/eks-anywhere: designs/auto-certs-renewal.md

# Auto Certificates Renewal

## Context

As a user who deployed K8s cluster with EKS-A, I want an option to automatically renew the certificates in etcd and control plane nodes before expiry in 1 year.

### External ETCD node certificates

* etcd node’s certificates are stored in /etc/etcd/pki (/var/lib/etcd/pki in bottlerocket)
* each certificate comes in pair: key and crt

> * all etcd nodes contain the same key pairs no matter if it’s the first node (not in term of contents)

Reviewer: ```suggestion
* all etcd nodes contain the same key pair names no matter if it’s the first node (not in term of contents)
```
?

In Stacked etcd nodes topology, etcd is running inside the cp node. 

We can renew all certificates by just redeploying cp nodes

1. Expose [rolloutBefore.certificatesExpiryDays](https://cluster-api.sigs.k8s.io/tasks/certs/auto-rotate-certificates-in-kcp.html) field from KubeadmControlPlane in EKSA cluster spec. Add pass the field to KubeadmControlPlane

### External etcd Nodes

In external etcd nodes topology, etcd is running in a separate machine besides cp machine. In addition to roll out cp nodes, we also need to roll out etcd nodes, with the following changes on top of above:

1. Add rolloutBefore.certificatesExpiryDays field to [EtcdadmCluster](https://github.com/aws/etcdadm-controller/blob/main/api/v1beta1/etcdadmcluster_types.go#L38), exposing and passing it from EKSA. The business logic of this field and reconciler should be:
    1. apply machine annotation `machine.cluster.x-k8s.io/certificates-expiry` during etcd node creation.
    2. check if roll out is needed. Refer https://github.com/kubernetes-sigs/cluster-api/blob/main/controlplane/kubeadm/internal/controllers/controller.go#L380.
    3. if roll out is needed for any etcd node, upgrade all etcd nodes.

> However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.

Author: This is still necessary. It is a bug existing today: a user would have apiserer-etcd-client expired even if he did the rollout for etcd and cp everyday.

The proposed solution by rollout etcd only would only make sure that cp doesn't have an extra rollout

> However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.

Reviewer 2: OK, help me walk through this pls, I might be missing something.
The etcd rollout to renew certs will always happen before the CP rollout. In fact, it's etcd the one that will cause the CO to rollout.

Then, why can't we just renew the apiserver-etcd-client secret just before the etcd rollout? That way it will be fresh when the CP nodes get rolled out.

> However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.

Author: If we just renew apiserver-etcd-client cert before etcd rollout, the apiserver-etcd-client cert of control plane will be renewed. However, the bug still exists: as a user, when I rollout cp nodes, I want all certs in the cp nodes get renewed.

> However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.

Reviewer 2: why is it a bug? There is no contract saying that the client cert always has to be fresh.

> However, the bug still exists: as a user, when I rollout cp nodes, I want all certs in the cp nodes get renewed.

Where is this requirement coming from?

> However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.

Author: I would say it's an implicit requirement based on common sense.

If a user upgrades his cluster everyday (not enabling auto certs renew), and he still get this cert expired in 1 year, he would have a great chance calling this behaviour "bug".

> However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.

Reviewer 3: I think we need to talk through the problem and solution some more. Are we able to schedule a call?

### Cluster Yaml Spec Example

```
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: Cluster
spec:

>   certificatesConfiguration:

Reviewer: what other certConfig will there be?

Author: Such as specifying ETCD/CP CA certs like Chris mentioned, or something like customizing cert's expiring time, algorithm, subject etc. But most likely, we would not need those fields in the near future, so just "thinking big" and make it extensible here.

    rolloutBeforeExpiry: 
      // whether to rollout etcd and cp nodes before certs expiring
      enabled: true/false (false by default, auto certs renew require opt-in) 
      // allow adding finer control in the future
```

### Tech Details

1. `certificatesConfiguration.rolloutBeforeExpiry.enabled` should be true when user run “eksctl anywhere generate clusterconfig”
2. When `rolloutBeforeExpiry` is enabled, the cp/etcd nodes are rollout when certs are expiring in **7** days

> 3. etcd certs renewal has a side-effect of renewing cp certs (etcd-rollout → etcd api change → cp rollout → cp certs renewal). We should only renew etcd certs to avoid unnecessary cp rollout.

Author: Added

> 3. etcd certs renewal has a side-effect of renewing cp certs (etcd-rollout → etcd api change → cp rollout → cp certs renewal). We should only renew etcd certs to avoid unnecessary cp rollout.

Reviewer 2: > ETCD cluster rollout should undergo the same process as upgrading cluster

Could we go over that process in the doc?

It is the same process than upgrades, but the same exact solution won't work. Upgrades are initiated by an external entity (the user). This decision to renew certs comes from withing the controllers (eg. there are no external API changes).

> 3. etcd certs renewal has a side-effect of renewing cp certs (etcd-rollout → etcd api change → cp rollout → cp certs renewal). We should only renew etcd certs to avoid unnecessary cp rollout.

Author: Right, it's not exactly the same process. The intended message to convey is in fact:
```
ETCD cluster rollout should be triggered similarly as it's triggered by upgrading etcd cluster
```

2. ca pair: the only pair which is not generated by etcdadm, it comes from cloud-init script.
3. etcdctl-etcd-client pair: generated by etcdadm during `etcdadm join <endpoint>`
4. peer pair: generated by etcdadm `etcdadm join <endpoint>`
5. server pair: generated by etcdadm `etcdadm join <endpoint>`

### CP node certificates

* all certificates are located in /etc/kubernetes/pki  (/var/lib/kubeadm/pki in bottlerocket)
* all cp nodes contains the same key pair names

list of all the certificate files:

1. apiserver-etcd-client pair: from cloud-init script, but the source is etcdadm controller 
2. ca pair: from cloud-init script
3. front-proxy-ca pair: from cloud-init script
4. sa pair: from cloud-init script
5. etcd/ca.crt: from cloud-init script
6. apiserver-kubelet-client pair: from kubeadm
7. apiserver pair: from kubeadm
8. front-proxy-client pair: from kubeadm

The above certificate files are described in the official doc: https://kubernetes.io/docs/setup/best-practices/certificates/#all-certificates

### Worker node certificates

1. just ca.crt in /etc/kubernetes/pki, which should be generated after `kubeadm join`

## Goal

> 1. Auto renew all the above mentioned certificates except etcd CA and kube CA, which last for 10 years.

Reviewer 3: Do we have an open issue for handling this also? Or, perhaps there's an easy win to address this at the same time?

Author: No such issue was created as far as I know.

Reviewer 3: Are we able to handle it at the same time. I know we want to address the first problem, but both problems seem similar.

Reviewer 4: is 10 years configurable? what happens if there is a need to roll the certs before then? what if they dont want 10 years.

In Stacked etcd nodes topology, etcd is running inside the cp node. 

We can renew all certificates by just redeploying cp nodes

1. Expose [rolloutBefore.certificatesExpiryDays](https://cluster-api.sigs.k8s.io/tasks/certs/auto-rotate-certificates-in-kcp.html) field from KubeadmControlPlane in EKSA cluster spec. Add pass the field to KubeadmControlPlane

### External etcd Nodes

In external etcd nodes topology, etcd is running in a separate machine besides cp machine. In addition to roll out cp nodes, we also need to roll out etcd nodes, with the following changes on top of above:

1. Add rolloutBefore.certificatesExpiryDays field to [EtcdadmCluster](https://github.com/aws/etcdadm-controller/blob/main/api/v1beta1/etcdadmcluster_types.go#L38), exposing and passing it from EKSA. The business logic of this field and reconciler should be:
    1. apply machine annotation `machine.cluster.x-k8s.io/certificates-expiry` during etcd node creation.
    2. check if roll out is needed. Refer https://github.com/kubernetes-sigs/cluster-api/blob/main/controlplane/kubeadm/internal/controllers/controller.go#L380.
    3. if roll out is needed for any etcd node, upgrade all etcd nodes.

> However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.

Reviewer 3: > However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. 

Could you clarify this. Are we making a statement about the existing behavior or something that's the result of implementing part of a solution?

In conjunction with the clarification, I think we need to clearly state the problem before mentioning any solutions. Does the final sentence elaborate further on the problem for example? Perhaps combine that with the problem description from the first sentence.

> However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.

Reviewer 3: > We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours

What is the expiry duration on the client certificate currently?

Author: 1 year

### Cluster Yaml Spec Example

```
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: Cluster
spec:

>   certificatesConfiguration:

Reviewer 3: Would we be open to just `certs` or `certificates`? This object is all about configuration and I don't think stating that explicitly in the name adds much.

Author: I think we should bias to use the same vocabulary as kubernetes' official documentation where certificates is prefered  :
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/

Reviewer 3: `certificates` is fine. I'm just thinking we should drop 'configuration'.

Reviewer 3: Thoughts @d8660091?

    rolloutBeforeExpiry: 
      // whether to rollout etcd and cp nodes before certs expiring
      enabled: true/false (false by default, auto certs renew require opt-in) 

>       // allow adding finer control in the future

Reviewer 3: What's the rationale for `rolloutBeforeExpiry` being a map vs a bool?

For example
```yaml
spec:
  certificates:
    rolloutBeforeExpiry: true
```

Also curious if, instead of a bool, we should use a time with [`time.Parse()`](https://pkg.go.dev/time#ParseDuration) semantics.

Author: It's a decision that we made with PM, the rationale is making it easier for the user to opt-in and flexible if we get feedbacks from users that they want to specify when to renew the certs

```

### Tech Details

> 1. `certificatesConfiguration.rolloutBeforeExpiry.enabled` should be true when user run “eksctl anywhere generate clusterconfig”

Reviewer 3: If the default is to enable, do we want the inverse? I.e. `disabled`?

Author: The default is still "false". We only make it true when the user generates clusterconfig.

Reviewer 3: How does this impact existing customers - particularly those that may not realize they need to enable this?

> 2. When `rolloutBeforeExpiry` is enabled, the cp/etcd nodes are rollout when certs are expiring in **7** days

Reviewer 3: ```suggestion
2. When `rolloutBeforeExpiry` is enabled the nodes are rolled 7 days before expiry.
```

3. etcd certs renewal has a side-effect of renewing cp certs (etcd-rollout → etcd api change → cp rollout → cp certs renewal). We should only renew etcd certs to avoid unnecessary cp rollout.
4. ETCD cluster rollout should undergo the same process as upgrading cluster, making sure that etcd cluster is accessible from cp, aka, no downtime for the k8s cluster.

## Alternative Solutions

### Cronjob

> The script run by the job would need to be tailored/tested for each provider/OS combinations. It’s error prone and hard to develop/maintain. Manually renew certs in bottlerocket is very tricky. Instead of running the binary locally, you have to pull the etcdadm/kubeadm image locally, mount folders, configuring container runtime, calling bottlerocket api etc to make it work. Besides, when there is bug or dependencies change, you would perhap need to test the fix against multiple environments.

Reviewer 3: These points are really interesting, thanks for noting them. To be able to think critically about the alternatives I think we need a little elaboration on the 'why' and a little more explanation on the what. Where does this CronJob run? Why is it error prone? Why is renewing certs in BottleRocket tricky?

Author: added

# Auto Certificates Renewal

## Context

> As a user who deployed K8s cluster with EKS-A, I want an option to automatically renew the certificates in etcd and control plane nodes before expiry in 1 year.

Reviewer 4: Hey @d8660091, the problem this is seeking to solve is a bit unclear to me. I'd like to see these questions answered.

- What problems are you solving?
- How do they impact users?

### Worker node certificates

1. just ca.crt in /etc/kubernetes/pki, which should be generated after `kubeadm join`

## Goal

1. Auto renew all the above mentioned certificates except etcd CA and kube CA, which last for 10 years.

### Observation

1. We can renew all etcd’s certs by simply **recreating the external etcd machine**. If doing manually, we just need to delete the auto generated certs and run `sudo etcdadm join phase certificates`
2. We can renew all cp’s certs besides apiserver-etcd-client by **recreating the cp machine.** As we can see from the following diagram that the apiserver-etcd-client is generated by Etcdadm Controller, so this cert cannot be renewed by just recreating the cp machine.
3. No need to do auto renew certs in worker nodes.

## Solution

Etcd can be deployed in [two topologies.](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/) Certs renew mechanism is slightly different in those two scenarios.

### Stacked etcd Nodes

In Stacked etcd nodes topology, etcd is running inside the cp node. 

> We can renew all certificates by just redeploying cp nodes

Reviewer 4: is this what we're telling customers to do? I'd like to see more details on the customer experience.

1. Expose [rolloutBefore.certificatesExpiryDays](https://cluster-api.sigs.k8s.io/tasks/certs/auto-rotate-certificates-in-kcp.html) field from KubeadmControlPlane in EKSA cluster spec. Add pass the field to KubeadmControlPlane

### External etcd Nodes

In external etcd nodes topology, etcd is running in a separate machine besides cp machine. In addition to roll out cp nodes, we also need to roll out etcd nodes, with the following changes on top of above:

1. Add rolloutBefore.certificatesExpiryDays field to [EtcdadmCluster](https://github.com/aws/etcdadm-controller/blob/main/api/v1beta1/etcdadmcluster_types.go#L38), exposing and passing it from EKSA. The business logic of this field and reconciler should be:

>     1. apply machine annotation `machine.cluster.x-k8s.io/certificates-expiry` during etcd node creation.

Reviewer 3: What happens if I remove the annotation?

### Worker node certificates

1. just ca.crt in /etc/kubernetes/pki, which should be generated after `kubeadm join`

## Goal

1. Auto renew all the above mentioned certificates except etcd CA and kube CA, which last for 10 years.

### Observation

1. We can renew all etcd’s certs by simply **recreating the external etcd machine**. If doing manually, we just need to delete the auto generated certs and run `sudo etcdadm join phase certificates`
2. We can renew all cp’s certs besides apiserver-etcd-client by **recreating the cp machine.** As we can see from the following diagram that the apiserver-etcd-client is generated by Etcdadm Controller, so this cert cannot be renewed by just recreating the cp machine.
3. No need to do auto renew certs in worker nodes.

## Solution

Etcd can be deployed in [two topologies.](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/) Certs renew mechanism is slightly different in those two scenarios.

### Stacked etcd Nodes

In Stacked etcd nodes topology, etcd is running inside the cp node. 

We can renew all certificates by just redeploying cp nodes

> 1. Expose [rolloutBefore.certificatesExpiryDays](https://cluster-api.sigs.k8s.io/tasks/certs/auto-rotate-certificates-in-kcp.html) field from KubeadmControlPlane in EKSA cluster spec. Add pass the field to KubeadmControlPlane

Reviewer 2: nit: we are not exposing `certificatesExpiryDays` directly in out spec. So maybe "control rolloutBefore.certificatesExpiryDays from EKS-A, setting a fixed value (7 days) when auto-renew certificates is enabled". Or something like that.
